server/selection_algorithm.go

```diff
@@ -0,0 +1,104 @@
+package server
+
+import (
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/golang/glog"
+	"math"
+	"math/rand"
+)
+
+type ProbabilitySelectionAlgorithm struct {
+	MinPerfScore float64
+
+	StakeWeight float64
+	PriceWeight float64
+	RandWeight  float64
+
+	PriceExpFactor float64
+}
+
+func (sa ProbabilitySelectionAlgorithm) Select(addrs []ethcommon.Address, stakes map[ethcommon.Address]int64, prices map[ethcommon.Address]float64, perfScores map[ethcommon.Address]float64) ethcommon.Address {
+	filtered := sa.filter(addrs, perfScores)
+	probabilities := sa.calculateProbabilities(filtered, stakes, prices)
+	return selectBy(probabilities)
+}
+
+func (sa ProbabilitySelectionAlgorithm) filter(addrs []ethcommon.Address, scores map[ethcommon.Address]float64) []ethcommon.Address {
```

Comment:
Has the price filtering already happened by this stage? It might be nice to bring that in here too

Reply:
No, filtering by price happens here: 
https://github.com/livepeer/go-livepeer/blob/f069a66779c238f64247f2cb2f666257c77fb4a4/discovery/db_discovery.go#L74

Hmm, actually it would be nice to move it into `filter()`. Let me try moving it.

Reply:
Ok, I gave it a second thought and it's not so easy to move it... so I'll not move it in this PR.

The problem is that it currently works like this:
- Select all Os that have price < max price
- From these Os, send `GetInfo()` to each
- Wait `500ms` for the replies (if no O replied, then wait `1s`, then wait `2s` and so on)
- Then from these Os, use the Selection Algorithm

Now if move this price check to the Selection Algorithm, then what may happen is that if we select the max price low, then all Os that replied in `500ms` may have price too high. That would result in a situation in which we don't select any Os, just because some of them have good price, but didn't reply in `500ms`, but others reply in `500ms` but have price too high.

Technically, it's the same issue with filtering by performance score, but performance score is highly correlated with replying fast to `GetOrch()`, so I'm not afraid about this case.

Reply:
Okay yeah, let's leave it for now then